Mage.Sets/src/mage/cards/c/CavernOfSouls.java

```diff
@@ -121,7 +120,7 @@ public boolean apply(Game game, Ability source, UUID originalId, Cost costToPay)
 
 class CavernOfSoulsWatcher extends Watcher {
 
-    private List<UUID> spells = new ArrayList<>();
+    private final Map<UUID, Integer> spells = new HashMap<>();
     private final UUID originalId;
 
     public CavernOfSoulsWatcher(UUID originalId) {
```

Comment:
There are another card with same effect and watcher code -- `BoseijuWhoSheltersAll`. It can be bugged too. I found not fixed link to that card in feedback report: #5155

> fix Bosejuwho shelters all: Daze was able to counter show and tell eventhough it was cast with Boseju (reported by Kwak675 / 2018-08-01 16:53:01.569).

You can try to fix it too.

Reply:
Boseiju got the same treatment